### PR TITLE
Feature/cyclic hash

### PR DESCRIPTION
**Pull request by wrightmhw**
Adding in Cyclic Hash for K > 32 as per issues #60.

### CODE REVIEW

**Comment by mr-c** [comment]
Code compiles and runs; only 10 errors and 198 failures when running the tests!

**Comment by mr-c** [comment]
Now 2 test errors and 206 failures; progress!

**Comment by ctb** [comment]
:)

File: lib/kmer_hash.cc [hunk 1]
```diff
@@ -31,23 +31,95 @@ HashIntoType _hash(const char * kmer, const WordLength k,
...
+}
 
-    _h = h;
-    _r = r;
+HashIntoType _revcyclichash(const char * kmer_string, const WordLength k)
+{
+    // Our list of random numbers for each base value
+    // Hard coding in random vals from array
+    // Random values generated from www.random.org
+    HashIntoType compliment_vals[] = {491989698, 127908739, 392233993, 303837760};
```
Comment (ctb) [inline lib/kmer_hash.cc]: complement, not compliment :)

File: tests/test_sandbox_scripts.py [hunk 2]
```diff
@@ -109,6 +109,10 @@ def test_sweep_reads():
 
     print os.listdir(in_dir)
 
+    assert os.path.exists(out1)
```
Comment (ctb) [inline tests/test_sandbox_scripts.py]: do these belong in a separate PR?

Comment (mr-c) [inline tests/test_sandbox_scripts.py]: I can pull them out separately if you want.

Comment (ctb) [inline tests/test_sandbox_scripts.py]: On Wed, Mar 11, 2015 at 03:52:14PM -0700, Michael R. Crusoe wrote:
> > @@ -109,6 +109,10 @@ def test_sweep_reads():
> >  
> >      print os.listdir(in_dir)
> >  
> > +    assert os.path.exists(out1)
> 
> I can pull them out separately if you want.

Just curious why they're part of this PR, which promises to be rather
far reaching - yes, might be better to split off the cleanup changes.

Comment (mr-c) [inline tests/test_sandbox_scripts.py]: Done: #889

**Comment by mr-c** [comment]
Some thoughts from an unsent email:

Currently we use a 2-bit DNA encoding scheme as the hash function. It has the advantages of being 'perfect', reversible, and collision free. Its disadvantage the output scales in size with the k-mer size and currently the codebase limits this to 64 bits, or a k-mer size of 32.

The cyclic hash has the advantage of a fixed size and its speed scales linearly with the k-mer size. However it is not reversible and collisions can (and will) happen.

Other hashing algorithms (like smhasher) may have a speed advantage but will also have the collision and irreversibility problem. And most won't be able to produce hashes for all the kmers in a string without fully hashing each kmer -- the 2bit kmerIterator uses a bit shift and I see how to write a version for the cyclic hash that also reuses the previous hash.

It is important to remind ourselves that we don't actually care about the size of the output of the hash function as it is used to index into the bloom filter/count-min sketch tables by taking the modules of the table sizes.

Therefore we could quadruple the size of the hash type which would support k-mer sizes up to 128 bases without increasing the memory requirements.

**Comment by ctb** [comment]
Additional note: we only care about reversibility for our sparse graph
implementation.  So if we allow multiple hash functions, selectable
at object creation, that would give us the flexibiltiy we need.

**Comment by mr-c** [comment]
@ctb Okay. What about the collision issue?

**Comment by ctb** [comment]
I think collisions only matter where we need reversibility. No?

**Comment by mr-c** [comment]
A collision at the hashing stage guarantees a collision at the bit setting or count incrementing stage regardless of bloom filter or count-min sketch table number or sizes. Frankly we could skip hashing and take the numerical value of the string and get the table locations that way.

**Comment by ctb** [comment]
On Tue, Mar 31, 2015 at 03:55:43AM -0700, Michael R. Crusoe wrote:
> A collision at the hashing stage guarantees a collision at the bit setting or count incrementing stage regardless of bloom filter or count-min sketch table number or sizes. Frankly we could skip hashing and take the numerical value of the string and get the table locations that way.

"Taking the numerical value of the string" *is* hashing, but it's non-random
and could lead to neighborhood structure.

I'm missing something here...? collisions at the hash level shouldn't matter
more than collisions at the CMS/Bloom filter stage, unless they're non-random.

**Comment by mr-c** [comment]
On Tue, Mar 31, 2015 at 1:58 PM C. Titus Brown <notifications@github.com>
wrote:

> On Tue, Mar 31, 2015 at 03:55:43AM -0700, Michael R. Crusoe wrote:
> > A collision at the hashing stage guarantees a collision at the bit
> setting or count incrementing stage regardless of bloom filter or count-min
> sketch table number or sizes. Frankly we could skip hashing and take the
> numerical value of the string and get the table locations that way.
>
> "Taking the numerical value of the string" *is* hashing, but it's
> non-random
> and could lead to neighborhood structure.
>

Yep. Though the 2-bit representation is also non-random.

I'm missing something here...? collisions at the hash level shouldn't matter
> more than collisions at the CMS/Bloom filter stage, unless they're
> non-random.
>

If the hashing function produces the same output for two inputs it will
cause the same location to be chosen in the Bloom filter or Count-min
sketch for both inputs in all tables regardless of the parameters chosen.

If a collision happens due to the size of the bloom filter or count-min
sketch then the best case scenario is that it will happen only in one of
the tables due to their different sizes and the modulus function. (The
worst case being that the table is so small and/or crowded that collisions
occur in every table with a different colliding k-mer).

**Comment by ctb** [comment]
On Tue, Mar 31, 2015 at 04:23:43AM -0700, Michael R. Crusoe wrote:
> On Tue, Mar 31, 2015 at 1:58 PM C. Titus Brown <notifications@github.com>
> wrote:
> 
> > On Tue, Mar 31, 2015 at 03:55:43AM -0700, Michael R. Crusoe wrote:
> > > A collision at the hashing stage guarantees a collision at the bit
> > setting or count incrementing stage regardless of bloom filter or count-min
> > sketch table number or sizes. Frankly we could skip hashing and take the
> > numerical value of the string and get the table locations that way.
> >
> > "Taking the numerical value of the string" *is* hashing, but it's
> > non-random
> > and could lead to neighborhood structure.
> >
> 
> Yep. Though the 2-bit representation is also non-random.

Yep.  Good reason to move away from it :)

> I'm missing something here...? collisions at the hash level shouldn't matter
> > more than collisions at the CMS/Bloom filter stage, unless they're
> > non-random.
> >
> 
> If the hashing function produces the same output for two inputs it will
> cause the same location to be chosen in the Bloom filter or Count-min
> sketch for both inputs in all tables regardless of the parameters chosen.
> 
> If a collision happens due to the size of the bloom filter or count-min
> sketch then the best case scenario is that it will happen only in one of
> the tables due to their different sizes and the modulus function. (The
> worst case being that the table is so small and/or crowded that collisions
> occur in every table with a different colliding k-mer).

Yes, but collisions from the hash function should be many, many orders of
magnitude less frequent than collisions from the limited size of the Bloom
filter.  For k=20, collisions from a random hash function will only
dominate if we have a hash table that's ~TB in size, right?

So unless there's structure in the hash function, there is no reason to
worry about this.

**Comment by mr-c** [comment]
But we aren't talking about k=20 we're talking about k=32..150

**Comment by ctb** [comment]
Hash collisions decrease or stay the same with increasing k, depending on what you're doing. For our applications they (at worse) stay the same.

But maybe I'm missing something. @camillescott, @luizirber?

--
Titus Brown, ctbrown@ucdavis.edu

**Comment by camillescott** [comment]
If you're worried about exact collisions between smhasher's 128-bit output and our 4^k output, the odds of smhasher's output even falling with the same range as our 2-bit rep are *astronomically small*... 5.4210109e-20 when K=32, to be precise. And this is an *extreme* upper bound, to say the very least (used out of laziness ;)).

In general, the odds of the direct collisions you describe are terribly terribly tiny. When we talk about collisions in hashtables, we're usually talking about the actual placement algorithm, not the conversion from string to integer. Simple modulus is probably the worst placement method, but that's part of the point of the bloom filter -- so long as there isn't collision prior to placement, it works as intended, and we already know that collision prior to placement is extremely unlikely.

**Comment by mr-c** [comment]
I worried about collisions between different k-mers using the cyclic hash
(64 bits) with a k-mer space of 4^k using k values up to 100 or 150.

**Comment by camillescott** [comment]
Ah, I see -- it was unclear what sort of collisions you were talking about before (I read it as collisions between different hashing methods, not collisions in the cyclic impl). Then yes, if cyclic hash has a >64-bit domain and a 64-bit codomain, it necessarily will have collisions. I believe @ctb is confusing increasing k-mer diversity within a sample, which reduces collisions if all hashtable parameters are held constant, with a random sampling of a larger k-mer space.

**Comment by ctb** [comment]
On Tue, Mar 31, 2015 at 09:31:46AM -0700, Camille Scott wrote:
> Ah, I see -- it was unclear what sort of collisions you were talking about before (I read it as collisions between different hashing methods, not collisions in the cyclic impl). Then yes, if cyclic hash has a >64-bit domain and a 64-bit codomain, it necessarily will have collisions. I believe @ctb is confusing increasing k-mer diversity within a sample, which reduces collisions if all hashtable parameters are held constant, with a random sampling of a larger k-mer space.

Update: talked with Camille, verified that we were on the same page, neither
of us sees a likely problem.

**Comment by mr-c** [comment]
Glad you two chatted, unclear as to what you both don't see a problem with: switching to longer 2bit encoding as the hash function or the use of collision generating 64bit hash functions with k>32?

**Comment by ctb** [comment]
Collisions, primarily. Reversibility is nice to have but not imo that high a priority for k > 32.

--
Titus Brown, ctbrown@ucdavis.edu

**Comment by mr-c** [comment]
*shrug* I'll continue with my multi-hash refactor.

**Comment by ctb** [comment]
(Can anyone tell me how #32 is relevant?)

**Comment by mr-c** [comment]
Can't see any relevance from where I sit, no.

**Comment by ctb** [comment]
Superseded by #1511 and related work.
